src/Symfony/Component/HttpClient/MiddlewareHttpClient.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Component\HttpClient\Middleware\MiddlewareInterface;
+use Symfony\Component\HttpClient\Middleware\MiddlewareStack;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+class MiddlewareHttpClient implements HttpClientInterface
+{
+    /**
+     * @var HttpClientInterface
+     */
+    private $httpClient;
+    /**
+     * @var MiddlewareStack
+     */
+    private $requestMiddlewareStack;
+
+    public function __construct(HttpClientInterface $httpClient, MiddlewareStack $requestMiddlewareStack)
```

Comment:
`MiddlewareStack` appears to be a fancy wrapper around an `array_reduce` call. Do we really need that class? Couldn't we just pass an array of middlewares instead?

Reply:
This is my only minor concern actually.

Reply:
we could. I just extracted this middleware logic to be able to replace it if needed and follow SRP
MiddlewareStack is here to create a middleware stack with certain logic. It can decide later to define priority middleware flag or whatever
MiddlewareHttpClient is an HTTPclient surrounded by middlewares

Reply:
> I just extracted this middleware logic to be able to replace it if needed and follow SRP

Well that's my point: There is no logic in that class. The main logic of how to assemble the class stack is in the builder function which is injected by the `MiddlewareHttpClient`. The responsibility of assembling the call stack is thus shared between the two classes. I would suggest to either

* inline the `array_reduce` call and drop `MiddlewareStack` or
* move the builder function into `MiddlewareStack`